coalib/bears/Bear.py

```diff
@@ -20,10 +20,29 @@
 from coalib.settings.FunctionMetadata import FunctionMetadata
 from coalib.settings.Section import Section
 from coalib.settings.ConfigurationGathering import get_config_directory
+from coalib.settings.Setting import typed_list
 
 from .meta import bearclass
 
 
+def _debugger_flag_setter(bear):
+    """
+    Set the value of `debugger` flag.
+
+    :param bear: Bear object
+    :return:     True if ``debug_bears`` is ``True`` or if bear name specified
+                 in ``debug_bears`` setting matched with the name of parameter
+                 bear.
+    """
+    if bear is None:
+        return False
+    if 'debug_bears' not in bear.section:
+        return False
+    return (bear.name.lower() in (
+            map(str.lower, typed_list(str)(bear.section['debug_bears']))) or
+            typed_list(bool)(bear.section['debug_bears'])[0])
```

Comment:
what is `typed_list(bool)(bear.section['debug_bears'])[0]` ?

that is buggy as.  I know what you are trying to do.  You want to hide the ValueError exceptions which could happen.  But it only hides some of them, because that is not what `typed_list(bool)` is designed to do.

You need to let go of that horrible `or` you have in the `return`.  It is forcing you to write more complicated and buggy code.

`bool(bear.section['debug_bears'])` can raise exceptions.  So, you need to catch them.  We even have a MR related to it at https://gitlab.com/coala/coala-utils/merge_requests/34/diffs .

Then after you have finished checking if it is a bool, then do another check if it is a list of bear names.
And do *not* use typed_list(str).  `bear.section['debug_bears']` is already the appropriate type.  Read the documentation about that type for how to use it correctly.

Reply:
I didn't get this part
>And do not use typed_list(str). bear.section['debug_bears'] is already the appropriate type

Can you tell me how am I going to make this --> 'ABear, Bbear' to a list without using `typed_list`?

Reply:
>bool(bear.section['debug_bears']) can raise exceptions. So, you need to catch them. 

doen't user know that they have passed wrong value (`ValueError`)?

>do another check if it is a list of bear names.

There are cases and here cases are many like let suppose user passed `ABear, BBear` and If I do something like `bool(bear.section['debug_bears'])` it will give `ValueError`

Reply:
After suggesting your changes this will be the code
```python
def _debugger_flag_setter(bear):
    """
    Set the value of `debugger` flag.

    :param bear: Bear object
    :return:     True if ``debug_bears`` is ``True`` or if bear name specified
                 in ``debug_bears`` setting matched with the name of parameter
                 bear.
    """
    if bear is None:
        return False
    if 'debug_bears' not in bear.section:
        return False
    try:
        return bool(bear.section['debug_bears'])
    except ValueError:
        return bear.name.lower() in (
                map(str.lower, typed_list(str)(bear.section['debug_bears'])))
```
And I still don't know why you said to not use `typed_list(str)`

Reply:
Read `Setting` and `StringConverter`; docs, code, whatever.

Do *not* use `typed_list(str)`!  It is that simple.  Any attempt to use that is not understanding what a `Setting` is, and is obfusicating the code!  It is almost like going `str(bool(None))`.

And for goodness sakes, never do complicated code in a exception handler.  This is bug heaven:
```py
    except ValueError:
        return bear.name.lower() in (
                map(str.lower, typed_list(str)(bear.section['debug_bears'])))
```
Code should tell a story.  Someone else should be able to read it and understand it almost immediately.  If they cant, it needs to be rewritten.

Please do not reply further.  You've been told what not to do, and to lay out your code better.  Submit a patch which does what was requested.  This isnt a debate.

Reply:
Sorry, does this code is acceptable?
```python
def _debugger_flag_value(bear):
    """
    Value of `debugger` flag.

    :param bear: Bear object
    :return:     True if ``debug_bears`` is ``True`` or if bear name specified
                 in ``debug_bears`` setting matched with the name of parameter
                 bear.
    """
    if bear is None:
        return False
    if 'debug_bears' not in bear.section:
        return False
    try:
        return bool(bear.section['debug_bears'])
    except ValueError:
        pass
    return bear.name.lower() in (
            map(str.lower, list(bear.section['debug_bears'])))

```

Reply:
im not reviewing code in comments that hasnt had CI check it and coverage fixed.